File: packages/validators/validate-arw.mjs
#!/usr/bin/env node
// Validate an ARW manifest (YAML) against a JSON Schema using Ajv
import fs from 'node:fs';
import path from 'node:path';
import process from 'node:process';
import { fileURLToPath } from 'node:url';
import Ajv from 'ajv';
import addFormats from 'ajv-formats';
import yaml from 'yaml';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const [, , manifestPath, schemaPath] = process.argv;
if (!manifestPath || !schemaPath) {
  console.error('usage: node tools/validators/validate-arw.mjs <manifest.yaml> <schema.json>');
  process.exit(2);
}

const manifestStr = fs.readFileSync(manifestPath, 'utf8');
const data = yaml.parse(manifestStr);

const schemaStr = fs.readFileSync(schemaPath, 'utf8');
const schema = JSON.parse(schemaStr);

const ajv = new Ajv({ allErrors: true, strict: false });
addFormats(ajv);
const validate = ajv.compile(schema);
const valid = validate(data);

if (!valid) {
  console.error('❌ Manifest failed validation:');
  for (const err of validate.errors) {
    const loc = err.instancePath || '(root)';
    console.error(`  - ${loc} ${err.message}`);
  }
  process.exit(1);
} else {
  console.log('✅ Manifest is valid.');
}

File: packages/validators/validate-arw.py
#!/usr/bin/env python3
import argparse, sys, json, yaml
from jsonschema import validate, Draft7Validator
from jsonschema.exceptions import ValidationError

def main():
    ap = argparse.ArgumentParser(description="Validate ARW /llms.txt against ARW JSON Schema")
    ap.add_argument("manifest", help="Path or URL to llms.txt/llms.yaml (YAML)")
    ap.add_argument("--schema", default="schemas/arw_model.json", help="Path to ARW JSON Schema")
    args = ap.parse_args()

    # Load schema
    with open(args.schema, "r") as f:
        schema = json.load(f)

    # Load YAML
    with open(args.manifest, "r") as f:
        data = yaml.safe_load(f)

    v = Draft7Validator(schema)
    errors = sorted(v.iter_errors(data), key=lambda e: e.path)
    if errors:
        print("❌ Manifest failed validation:")
        for e in errors:
            path = "/".join([str(p) for p in e.path])
            print(f"  - {path}: {e.message}")
        sys.exit(1)
    else:
        print("✅ Manifest is valid.")
        sys.exit(0)

if __name__ == "__main__":
    main()
